### DISCUSSION TITLE
How to convert ArgumentParser code to Mojo?

### DISCUSSION THREAD

**Discussion by huiliukeith-mineral** [modular/modular#1029] [General]
I am trying to convert code like this to Mojo
```
    parser = argparse.ArgumentParser()
    parser.add_argument('--weights', type=str, default='yolo7.pt', help='initial weights path')
    parser.add_argument('--cfg', type=str, default='', help='model.yaml path')
    parser.add_argument('--data', type=str, default='data/coco.yaml', help='data.yaml path')
```

and what I did
```
let argparse = Python.import_module("argparse")
let parser = argparse.ArgumentParser()
    parser.add_argument('--weights', type=str, default='yolo7.pt', help='initial weights path')
    parser.add_argument('--cfg', type=str, default='', help='model.yaml path')
    parser.add_argument('--data', type=str, default='data/coco.yaml', help='data.yaml path')
```

and I got 
```
yolov7/train.mojo:566:43: error: use of unknown declaration 'str', 'fn' declarations require explicit variable declarations"
```

**Comment by pu2is** [comment]
extactly same exception :(

**Comment by pu2is** [comment]
> I am trying to convert code like this to Mojo
> 
> ```
>     parser = argparse.ArgumentParser()
>     parser.add_argument('--weights', type=str, default='yolo7.pt', help='initial weights path')
>     parser.add_argument('--cfg', type=str, default='', help='model.yaml path')
>     parser.add_argument('--data', type=str, default='data/coco.yaml', help='data.yaml path')
> ```
> 
> and what I did
> 
> ```
> let argparse = Python.import_module("argparse")
> let parser = argparse.ArgumentParser()
>     parser.add_argument('--weights', type=str, default='yolo7.pt', help='initial weights path')
>     parser.add_argument('--cfg', type=str, default='', help='model.yaml path')
>     parser.add_argument('--data', type=str, default='data/coco.yaml', help='data.yaml path')
> ```
> 
> and I got
> 
> ```
> yolov7/train.mojo:566:43: error: use of unknown declaration 'str', 'fn' declarations require explicit variable declarations"
> ```

``` mojo
from python import Python

fn main() raises:
    let argparse = Python.import_module("argparse")
    let parser = argparse.ArgumentParser()
    parser.description = "your description here"
    let example_arg = parser.add_argument("-d", "--database")
    example_arg.require = True
    example_arg.help = "help text"
    print(parser)
    print(example_arg)
```

the output is

```
ArgumentParser(prog='', usage=None, description='your description here', ...)
_StoreAction(option_strings=['-d', '--database'], dest='database', nargs=None, ..., help='help text')
```
but if i append following code there is exception:
```
...
print(...)
let args = parser.parse_args()
```

execution with: `mojo main.mojo -d foo`
got exception:

```
usage: [-h] -d DATABASE
: error:  the following arguments are required: -d/--database
unhandled exception caught during execution: 2
mojo: error: execution exited with a non-zero result: 1
```

then i changed the `example_arg.required = False` and add `example_arg.default = "foo_1"`
the exception is no longer exists.
if i print the `args.database` the output is `foo_1`, which is correct.

the problem could be: how to set required var through a command line

**Comment by gryznar** [comment]
Unknown declaration `str` comes from the fact that Mojo does not support Python's `str` yet. Try to use `str` from builtins:
```
let builtins = Python.import_module("builtins")
parser.add_argument('--weights', type=builtins.str, default='yolo7.pt', help='initial weights path')
```
